pkg/distributors/contracts/MultiRewards.sol

```diff
@@ -70,191 +75,300 @@ contract MultiRewards is IMultiRewards, IDistributor, ReentrancyGuard, Temporari
+    ) public nonReentrant whenNotPaused updateReward(pool, receiver) {
         require(amount > 0, "Cannot stake 0");
-        _totalSupply[stakingToken] = _totalSupply[stakingToken].add(amount);
-        _balances[stakingToken][receiver] = _balances[stakingToken][receiver].add(amount);
-        stakingToken.safeTransferFrom(msg.sender, address(this), amount);
-        emit Staked(address(stakingToken), receiver, amount);
+        _totalSupply[pool] = _totalSupply[pool].add(amount);
+        _balances[pool][receiver] = _balances[pool][receiver].add(amount);
+        pool.safeTransferFrom(msg.sender, address(this), amount);
+        emit Staked(address(pool), receiver, amount);
     }
 
     /**
      * @notice Stake tokens using a permit signature for approval
+     * @param pool The bpt being staked to earn rewards
      * @param amount    Amount of allowance
      * @param deadline  The time at which this expires (unix time)
      * @param v         v of the signature
      * @param r         r of the signature
      * @param s         s of the signature
      */
     function stakeWithPermit(
-        IERC20 stakingToken,
+        IERC20 pool,
         uint256 amount,
         uint256 deadline,
+        address recipient,
         uint8 v,
         bytes32 r,
         bytes32 s
     ) public {
-        IERC20Permit(address(stakingToken)).permit(msg.sender, address(this), amount, deadline, v, r, s);
-        stake(stakingToken, amount, msg.sender);
+        IERC20Permit(address(pool)).permit(msg.sender, address(this), amount, deadline, v, r, s);
+        stake(pool, amount, recipient);
     }
 
-    function withdraw(IERC20 stakingToken, uint256 amount) public nonReentrant updateReward(stakingToken, msg.sender) {
+    function unstake(IERC20 pool, uint256 amount) public nonReentrant updateReward(pool, msg.sender) {
         require(amount > 0, "Cannot withdraw 0");
-        _totalSupply[stakingToken] = _totalSupply[stakingToken].sub(amount);
-        _balances[stakingToken][msg.sender] = _balances[stakingToken][msg.sender].sub(amount);
-        stakingToken.safeTransfer(msg.sender, amount);
-        emit Withdrawn(address(stakingToken), msg.sender, amount);
+        _totalSupply[pool] = _totalSupply[pool].sub(amount);
+        _balances[pool][msg.sender] = _balances[pool][msg.sender].sub(amount);
+        pool.safeTransfer(msg.sender, amount);
+        emit Withdrawn(address(pool), msg.sender, amount);
     }
 
-    function getReward(IERC20[] calldata stakingTokens) public nonReentrant {
-        _getReward(stakingTokens, false);
+    // todo accept array of claims [{pool, rewardToken}]
+    function getReward(IERC20[] calldata pools) public nonReentrant {
+        _getReward(pools, false);
     }
 
-    function getRewardAsInternalBalance(IERC20[] calldata stakingTokens) public nonReentrant {
-        _getReward(stakingTokens, true);
+    function getRewardAsInternalBalance(IERC20[] calldata pools) public nonReentrant {
```

Comment:
We should add a recipient here for implementing higher level reinvestment/swap/etc